### PR TITLE
documentation #765: added documentation to user endpoints

### LINKED ISSUE #765: Update Swagger doc for User (1)
We are updating the documentation for all endpoints and adding more _description and details_. 

One of the main controllers is User. You should update the documentation according to the **pattern** below:
```ts
  @ApiBearerAuth()
  @ApiOkResponse({
    type: DisciplineTeachersResponse,
  })
  @ApiBadRequestResponse({
    description: `\n
    InvalidQueryException:
      Type of discipline must be a field of enum
    InvalidDisciplineIdException:
      Discipline with such id is not found`,
  })
  @ApiUnauthorizedResponse({
    description: `\n
    UnauthorizedException:
      Unauthorized`,
  })
  @ApiForbiddenResponse({
    description: `\n
    NoPermissionException:
      You do not have permission to perform this action`,
  })
  @ApiParam({
    name: 'disciplineId',
    required: true,
    description: 'Id of certain discipline',
  })
  @ApiQuery({
    name: 'disciplineType',
    required: true,
    enum: DisciplineTypeEnum,
    description: 'Discipline type of some discipline',
  })
  @Access(PERMISSION.GROUPS_$GROUPID_DISCIPLINES_TEACHERS_GET, GroupByDisciplineGuard)
  @Get('/:disciplineId/teachers')
  async getAllByDiscipline (
    @Param('disciplineId', DisciplineByIdPipe) disciplineId: string,
    @Query('disciplineType') disciplineType: DisciplineTypeEnum
  ) {
    const teachers = await this.disciplineService.getTeachers(disciplineId, disciplineType);
    return { teachers };
  }
```
**Structure:**
- auth (if needed)
- ok response
- client errors
- params/query description
- access (if needed)

You should also add **field description** in the bady or response.
Add **summary** to endpoints (fix this issue).
Set **default values** for the optional fields

image

_My DM is always open for you. Happy coding_ 🐸

[Our Swagger](https://apidev.fictadvisor.com/api)
[Doc](https://docs.nestjs.com/openapi/introduction)

### PR DESCRIPTION
**Pull request by ArtemLyah**
- added documentation to verify
- added documentation to verifySuperhero
- added documentation to requestNewGroup
- added documentation to createSuperhero
- added documentation to getSelective
- added documentation to getSelectiveBySemesters

closes #765

### CODE REVIEW

File: src/v2/api/controllers/UserController.ts [hunk 1]
```diff
@@ -64,7 +66,36 @@ export class UserController {
...
+    description: `\n
+    NoPermissionException:
+      You do not have permission to perform this action`,
+  })
+  @ApiParam({
+    name: 'userId',
+    required: true,
+    description: 'Id of a user to verify',
+  })
+  @Access(PERMISSION.USERS_$USERID_VERIFY_STUDENT, TelegramGuard)
```
Comment (fokaaas) [inline src/v2/api/controllers/UserController.ts:98]: remove permission

File: src/v2/api/controllers/UserController.ts [hunk 2]
```diff
@@ -73,17 +104,38 @@ export class UserController {
...
+    NoPermissionException:
+      You do not have permission to perform this action`,
+  })
+  @ApiParam({
+    name: 'userId',
+    required: true,
+    description: 'Id of a superhero to verify',
+  })
+  @Access(PERMISSION.USERS_$USERID_VERIFY_SUPERHERO, TelegramGuard)
   @Patch('/:userId/verifySuperhero')
```
Comment (fokaaas) [inline src/v2/api/controllers/UserController.ts:131]: add @ApiUnauthorizedResponse

File: src/v2/api/controllers/UserController.ts [hunk 3]
```diff
@@ -73,17 +104,38 @@ export class UserController {
...
+    description: `\n
+    NoPermissionException:
+      You do not have permission to perform this action`,
+  })
+  @ApiParam({
+    name: 'userId',
+    required: true,
+    description: 'Id of a superhero to verify',
+  })
+  @Access(PERMISSION.USERS_$USERID_VERIFY_SUPERHERO, TelegramGuard)
```
Comment (fokaaas) [inline src/v2/api/controllers/UserController.ts:130]: remove permission

File: src/v2/api/controllers/UserController.ts [hunk 4]
```diff
@@ -113,13 +165,40 @@ export class UserController {
     ForbiddenException:
       Forbidden`,
   })
+  @ApiParam({
+    name: 'userId',
+    required: true,
+    description: 'Id of a user to request to the group',
+  })
+  @Access(PERMISSION.USERS_$USERID_GROUP_REQUEST)
+  @Patch('/:userId/requestNewGroup')
```
Comment (fokaaas) [inline src/v2/api/controllers/UserController.ts:174]: add @ApiUnauthorizedResponse

File: src/v2/api/controllers/UserController.ts [hunk 5]
```diff
@@ -113,13 +165,40 @@ export class UserController {
...
   requestNewGroup (
     @Param('userId', UserByIdPipe) userId: string,
     @Body() body: GroupRequestDTO,
   ) {
     return this.userService.requestNewGroup(userId, body);
   }
 
+  @ApiBearerAuth()
+  @ApiOkResponse()
```
Comment (fokaaas) [inline src/v2/api/controllers/UserController.ts:183]: where?

File: src/v2/api/controllers/UserController.ts [hunk 6]
```diff
@@ -113,13 +165,40 @@ export class UserController {
...
+    NoPermissionException:
+      You do not have permission to perform this action`,
+  })
+  @ApiParam({
+    name: 'userId',
+    required: true,
+    description: 'Id of a user which is going to be a superhero',
+  })
   @Access(PERMISSION.USERS_$USERID_SUPERHERO_CREATE)
   @Post('/:userId/superhero')
```
Comment (fokaaas) [inline src/v2/api/controllers/UserController.ts:203]: add @ApiUnauthorizedResponse

File: src/v2/api/controllers/UserController.ts [hunk 7]
```diff
@@ -129,6 +208,25 @@ export class UserController {
...
+    NoPermissionException:
+      You do not have permission to perform this action`,
+  })
+  @ApiParam({
+    name: 'userId',
+    required: true,
+    description: 'Id of a user',
+  })
   @Access(PERMISSION.USERS_$USERID_SELECTIVE_GET)
   @Get('/:userId/selective')
```
Comment (fokaaas) [inline src/v2/api/controllers/UserController.ts:231]: add @ApiUnauthorizedResponse

File: src/v2/api/controllers/UserController.ts [hunk 8]
```diff
@@ -138,19 +236,27 @@ export class UserController {
...
+    NoPermissionException:
+      You do not have permission to perform this action`,
   })
+  @ApiParam({
+    name: 'userId',
+    required: true,
+    description: 'Id of a user',
+  })
+  @Access(PERMISSION.USERS_$USERID_SELECTIVE_GET)
+  @Get('/:userId/selectiveBySemesters')
```
Comment (fokaaas) [inline src/v2/api/controllers/UserController.ts:259]: add @ApiUnauthorizedResponse

File: src/v2/api/dtos/ApproveDTO.ts [hunk 9]
```diff
@@ -6,15 +6,16 @@ import { ApiProperty } from '@nestjs/swagger';
...
     state: State;
 }
 
-export class ApproveStudentByTelegramDTO {
-  @IsEnum(State, validationOptionsMsg('State is not an enum'))
-    state: State;
-
+export class ApproveStudentByTelegramDTO extends ApproveDTO {
+  @ApiProperty({
+    description: 'whether the student is captain or not',
```
Comment (fokaaas) [inline src/v2/api/dtos/ApproveDTO.ts:17]: Capitalize

File: src/v2/api/dtos/ApproveDTO.ts [hunk 10]
```diff
@@ -6,15 +6,16 @@ import { ApiProperty } from '@nestjs/swagger';
 export class ApproveDTO {
   @ApiProperty({
     enum: State,
+    description: 'state for student verification',
```
Comment (fokaaas) [inline src/v2/api/dtos/ApproveDTO.ts:9]: Capitalize

File: src/v2/api/dtos/UpdateSuperheroDTO.ts [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+  @IsOptional()
+  @IsBoolean(validationOptionsMsg('Dorm is not a boolean'))
+    dorm?: boolean;
+  
+  @ApiPropertyOptional({
+    description: 'State for superhero verification',
+  })
+  @IsOptional()
+  @IsEnum(State, validationOptionsMsg('State is not an enum'))
+    state?: State;
```
Comment (fokaaas) [inline src/v2/api/dtos/UpdateSuperheroDTO.ts:19]: add default value

File: src/v2/api/responses/GroupResponse.ts [hunk 12]
```diff
@@ -3,33 +3,41 @@ import { State, RoleName } from '@prisma/client';
...
+  @ApiProperty({
+    description: 'Group code string',
+  })
     code: string;
 }
 
 export class ExtendedGroupResponse extends GroupResponse {
   @ApiProperty({
     enum: State,
+    description: 'State for the student in group: APPROVED, PENDING, DECLINED',
```
Comment (fokaaas) [inline src/v2/api/responses/GroupResponse.ts:20]: don't specify values in the description, enum does it

File: src/v2/api/responses/StudentResponse.ts [hunk 13]
```diff
@@ -2,60 +2,94 @@ import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
...
-  @ApiPropertyOptional({
+  @ApiProperty({
     type: ExtendedGroupResponse,
+    description: 'Group where the student is in',
   })
     group: ExtendedGroupResponse;
+  
+  @ApiProperty({
+    enum: State,
+    description: 'State for the student: APPROVED, PENDING, DECLINED',
```
Comment (fokaaas) [inline src/v2/api/responses/StudentResponse.ts:84]: don't specify values in the description, enum does it

File: src/v2/security/PERMISSION.ts [hunk 14]
```diff
@@ -62,6 +62,8 @@ export enum PERMISSION {
   USERS_UPDATE = 'users.update',
   USERS_DELETE = 'users.delete',
   USERS_GROUPS_SWITCH = 'users.groups.switch',
+  USERS_$USERID_VERIFY_STUDENT = 'users.$userId.verifyStudent',
+  USERS_$USERID_VERIFY_SUPERHERO = 'users.$userId.verifySuperhero',
```
Comment (fokaaas) [inline src/v2/security/PERMISSION.ts:66]: remove it

**Review by fokaaas — CHANGES_REQUESTED** [review]
You also need to add summary to each endpoint, but so that it doesn't erase description

**Review by fokaaas — APPROVED** [review]

File: src/v2/api/controllers/UserController.ts [hunk 15]
```diff
@@ -113,23 +188,91 @@ export class UserController {
     ForbiddenException:
       Forbidden`,
   })
+  @ApiParam({
+    name: 'userId',
+    required: true,
+    description: 'Id of a user to request to the group',
+  })
+  @ApiEndpoint({
+    summary: 'Verify student to be a superhero',
```
Comment (Ksrebniukk) [inline src/v2/api/controllers/UserController.ts:197]: the query description is the same as in the previous one

File: src/v2/api/controllers/UserController.ts [hunk 16]
```diff
@@ -73,17 +114,46 @@ export class UserController {
     return this.userService.verifyStudent(userId, isCaptain, state);
   }
 
-  @UseGuards(TelegramGuard)
+  @ApiBearerAuth()
+  @ApiOkResponse({
+    type: StudentsResponse,
```
Comment (Ksrebniukk) [inline src/v2/api/controllers/UserController.ts:119]: This is wrong. Print the another answer
image

**Review by Ksrebniukk — APPROVED** [review]

**Pull request merged**